libs/ballot-interpreter/README.md

```diff
@@ -183,23 +203,78 @@ etc.
 
 ### Score Bubble Marks
 
-Bubble marks are scored to determine which bubbles are filled in. Locating the
-bubble at `(x, y)` first locates the `y`th timing mark on the left and right
-sides of the ballot. Given a line segment between the centers of these two
-timing marks, the center of the bubble is presumed to be on this line `x / N`
-percent of the way from the left timing mark to the right timing mark, where `N`
-is the number of columns of timing marks on the ballot. The bubble is then
-scored by comparing a template bubble image to the actual contents of the scan
-at the bubble location. A score is computed by computing the number of new dark
-pixels compared to the template bubble image. The score is later compared to a
-threshold to determine if the bubble is filled in, but the core function simply
+Bubble marks are scored to determine which bubbles are filled in. The scoring
+process ([scoring.rs](src/bubble-ballot-rust/scoring.rs)) consists of two main
+steps: template matching and fill scoring.
+
+#### Bubble Localization
+
+To locate a bubble at grid coordinates `(column, row)`:
+
+1. Find the `row`th timing mark on the left and right sides of the ballot. If
+   `row` is fractional, interpolate vertically between the closest two rows.
+2. Account for timing marks being cropped during scanning or border removal by
+   adjusting the timing mark positions to use the expected width.
+3. Compute a line segment between the centers of the left and right timing marks.
+4. The expected bubble center is at position `column / (N - 1)` along this
+   segment, where `N` is the total number of columns in the grid.
+
+#### Template Matching
+
+To account for stretching and other distortions in the scanned image, the
+algorithm performs template matching within a search area:
```

Comment:
Maybe give some context on what the template is that's being matched?

Reply:
Linked to the scanned bubble template file.